File: server/qywk-cloud/qywk-common/qywk-common-core/src/main/java/com/qywk/common/core/utils/PageUtils.java
package com.qywk.common.core.utils;

import java.util.List;

/**
 * @author qlh
 * @date 2024/04/02 17:11
 * @description 分页工具类
 */
public class PageUtils<T> {

    private Integer pageNum;
    private Integer pageSize;
    private final List<T> src;

    public PageUtils(List<T> src, Integer pageNum, Integer pageSize){
        this.src = src;
        this.pageNum = pageNum;
        this.pageSize = pageSize;
    }

    public List<T> getRecords(){
        int total = src.size();
        int startIndex = (pageNum - 1) * pageSize;
        int endIndex = Math.min(startIndex + pageSize, total);
        if(startIndex > endIndex) return null;
        return src.subList(startIndex, endIndex);
    }

    public Integer getTotal(){
        return this.src.size();
    }

    public void setPageNum(Integer pageNum){
        this.pageNum = pageNum;
    }

    public void setPageSize(Integer pageSize){
        this.pageSize = pageSize;
    }

}

File: server/qywk-server/pkg/pageutils/pageutils.go
package pageutils

type PageUtils[T any] struct {
	PageNum  int
	PageSize int
	Src      []T
}

func NewPageUtils[T any](src []T, pageNum, pageSize int) *PageUtils[T] {
	return &PageUtils[T]{
		Src:      src,
		PageNum:  pageNum,
		PageSize: pageSize,
	}
}

func (p *PageUtils[T]) GetRecords() []T {
	total := len(p.Src)
	startIndex := (p.PageNum - 1) * p.PageSize
	endIndex := startIndex + p.PageSize
	if startIndex > total {
		return nil
	}
	if endIndex > total {
		endIndex = total
	}
	return p.Src[startIndex:endIndex]
}

func (p *PageUtils[T]) GetTotal() int {
	return len(p.Src)
}

func (p *PageUtils[T]) SetPageNum(pageNum int) {
	p.PageNum = pageNum
}

func (p *PageUtils[T]) SetPageSize(pageSize int) {
	p.PageSize = pageSize
}
